[Feature] Add $screen helper

muzafferdede opened the pull request:
Provides $screen helper for reactive breakpoints. 

~- Supports switchable Tailwind, Bootstrap, Bumla, Materialize grid breakpoints~
- Supports direct defined breakpoints as `$screen(1024)`

~It accepts 2 arguments~

**Usage**
`$screen(breakpoint/touch/orientation)`

By Default it uses Tailwind CSS
`$screen('lg')`
`$screen('md')`
...

~To use Bulma for example, we can define using framework slug/short name as second paramenter~
~`$screen('tablet','bl')` //Bulma~
....

We can also use direct screen sizes as:
`$screen(768)`

KevinBatdorf commented:
I think some of the code formatting is off compared to the rest of the codebase. I would think the linter would fix that though? @SimoTod is the config broken maybe? I remember your last push had a few extra semicolons too. For me it's removing them on save. Or is that my personal config working perhaps?

We probably should have had a [discussion](https://github.com/alpine-collective/alpine-magic-helpers/discussions) on this first before implementing it to figure out the api a bit more.

I think adding the breakpoints into the code feels wrong to me. Instead it might be better to let the user pass them in via another magic property inside `x-init`, like `$breakpoints({ 'lg': 1234, })`? Let's see what the others here have to say. 

Otherwise I think it will be pretty useful.

Would this also work adding a callback too?

`<div x-show="$screen('xl', v => console.log(v))"></div>` which would return a boolean but also fire a callback? Anyone have thoughts on that?

muzafferdede commented:
> I think some of the code formatting is off compared to the rest of the codebase. I would think the linter would fix that though? @SimoTod is the config broken maybe? I remember your last push had a few extra semicolons too. For me it's removing them on save. Or is that my personal config working perhaps?

Yeah i think i got confused as well, i disabled my "format on save" tho.

> We probably should have had a [discussion](https://github.com/alpine-collective/alpine-magic-helpers/discussions) on this first before implementing it to figure out the api a bit more.

The current api is inspired by [vue-screen](https://www.npmjs.com/package/vue-screen) but yeah, we can have a discussion. 

> I think adding the breakpoints into the code feels wrong to me. Instead it might be better to let the user pass them in via another magic property inside `x-init`, like `$breakpoints({ 'lg': 1234, })`? Let's see what the others here have to say.

I don't think it will make sense to provide breakpoints to each components, That's why i included it into the code. And these breakpoints are commonly used specially Tailwind grid. 

> Would this also work adding a callback too?

> `<div x-show="$screen('xl', v => console.log(v))"></div>` which would return a boolean but also fire a callback? Anyone have thoughts on that?

yeah, that could be easily implemented.

SimoTod commented:
> I think some of the code formatting is off compared to the rest of the codebase. I would think the linter would fix that though? @SimoTod is the config broken maybe? I remember your last push had a few extra semicolons too. For me it's removing them on save. Or is that my personal config working perhaps?

@KevinBatdorf This PR looks okay. The src folder doesn't have semicolon as expected. Which bit seems off to you?

I agree with Kevin, let's have a chat on the best path to follow first -> https://github.com/alpine-collective/alpine-magic-helpers/discussions/37

KevinBatdorf commented:
> The src folder doesn't have semicolon as expected. Which bit seems off to you?

Looks like I was looking at the `dist` code. Hadn't had my morning coffee yet I guess!

SimoTod commented:
@muzafferdede #40 gets merged in, you should be able to do
`import config from './config'`at the top and, in your helper, `config.get('breakpoints')` to fetch the current configuration (defaulted to tailwind).
If you want to document other frameworks in the readme, the way to set the custom config is 
<script>window.AlpineMagicHelpersConfig = { mobile: 0, tablet: 769, desktop: 1024, widescreen: 1216, fullhd: 1408 }</script> in your head section (either before or after the helper).

muzafferdede commented:
This is great, thanks. I was ready to create tests for overriding as well but seems you have handled that. I can handle documenting custom breakpoints for the other frameworks. 

BTW, shouldn't it be `window.AlpineMagicHelpersConfig.breakpoints = {mobile: .....}` ? We don't want overriding entire config. 

And how about if i want to override or extend breakpoints in $screen helper am i able to do so? I am thinking to add `$screen('touch')` and $screen('landscape')` keys to even extend the functionality. Not sure if after everything kick-in are we able to change things.

SimoTod commented:
Yeah you are correct, I missed `breakpoints` in the example. 
Either what you said or
`window.AlpineMagicHelpersConfig = { breakpoints: { mobile: 0, tablet: 769, desktop: 1024, widescreen: 1216, fullhd: 1408 } }`, it won't override the whole config (I'll add a test tomorrow).

About the second question, How would you implement `landscape` or `touch` in your helper? Maybe the config name should be generic and it should contain media queries rather then integers

muzafferdede commented:
> About the second question, How would you implement landscape or touch in your helper?


```js
                const screen = {
                    ...config.get('breakpoints'),
                    touch: 'ontouchstart' in window,
                    portrait: height > width,
                    landscape: width > height
                }
```

i already finish the $screen helper, now working on tests.

I have also tried to use matchMedia for `landscape` and `portrait` to check but for test i have to mock matchMedia method which didn't work properly. 

i am trying to achieve this:
```js
                const screen = {
                    ...config.get('breakpoints'),
                    touch: 'ontouchstart' in window,
                    portrait: window.matchMedia("(orientation: portrait)").matches,
                    landscape: window.matchMedia("(orientation: landscape)").matches
                }
```
Works on browser but i couldn't make it work on jest test. Specially don't know how to test `touch`. Maybe i just push the code so far and see if we can improve it.

SimoTod commented:
It makes sense. I don't think jest supports anything related to the screen since it doesn't run in a browser.

HugoDF commented:
> It makes sense. I don't think jest supports anything related to the screen since it doesn't run in a browser.

Doesn't Jest run with JSDOM by default? I think you can `Object.definedProperty(window, 'height', { value: 100 })` or whatever you need

SimoTod commented:
He's already testing that part. The question was about the touch check, we can add the relevant property used in the implementation to window but it would add little value as the test would just assert that the implementation uses the right property. I guess it's the best we can do here.

muzafferdede commented:
Alright, i have added a touch test which is basically a real life case where `window.ontouchstart` is in window. I simulate it basically assigning `window.ontouchstart = true` and test passes. As @SimoTod said, this is best we can do. 

I removed the config.js file from this pr since @SimoTod already implemented it. So keep in mind that test will passed once #40 merged.

SimoTod commented on src/screen.js:
```diff
@@ -0,0 +1,57 @@
+                }, 150)
+            }
+
+            window.addEventListener('resize', updateScreen)
```
There is a memory leak here. The helpers adds a new listener everytime screen gets evaluated and each listener triggers it's own component refresh so after a few interactions, there are loads of event listener attached to windows. This will be likely to make pages slow on a long run. you can add a console log inside the settimeout and see that the amount of log doubles for each alpine refresh

muzafferdede replied:
Yes!! That's is what i found as well. Function call multiplied on each resize event. I even tried our good friend `once` method. I thought it's a known behavior so i forget to mention it. What would be the best path here?

SimoTod replied:
You'll need to keep track of it somehow so you don't attach multiple listeners for the same component. It will kill performance quickly on big pages.

muzafferdede replied:
Not sure if it's proper way but i handled it by assigning a global variable on first init and check that to re-assign resize event.
```js
            let update
            const updateScreen = () => {
                clearTimeout(update)
                update = setTimeout(() => {
                    $el.__x.updateElements($el)
                }, 150)

                // set $screenHelperInitialized to prevent multiple calls
                window.$screenHelperInitialized = true
            }

            // bind resize event to window if not initialized
            if (!window.$screenHelperInitialized) {
                window.addEventListener('resize', updateScreen)
            }
```

SimoTod replied:
I'm not 100% sure (I haven't tried yet) but I feel like it won't work if you have 2 Alpine components using the helper.
For example
```html
<div x-data>
    <span x-show="$screen('lg')">Foo</span>
</div>
<div x-data>
    <span x-show="$screen('lg')">Bar</span>
</div>
```
I suspect only the first one will update.

SimoTod replied:
I might be wrong though 😂
Can you just double check pls?

muzafferdede replied:
Yeah, i think i need to dive in deeper level for this part. But i think i got an idea. Let me try today or tomorrow.

HugoDF commented on src/screen.js:
```diff
@@ -11,9 +11,15 @@ const AlpineScreenMagicMethod = {
                 }, 150)
+
+                // set $screenHelperInitialized to prevent multiple calls
+                window.$screenHelperInitialized = true
```
does it need to be on the window? can't we just have a variable on the closure?

muzafferdede replied:
We use `$el.__x.updateElements($el)` where it re-runs $screen helper each time windows resized. Which means everything starts from 0. I am not sure if this is expected behavior tho. All i am trying to do is observe screen width and update the breakpoints.

HugoDF replied:
I mean that this function/module (screen.js) has its own scope, we could store the listener/whether it's been initialised in the module instead of putting it on the window

muzafferdede replied:
hmm, you mean something like this?

```js
let Initialized = false;

const AlpineScreenMagicMethod = {
    start() {
        Alpine.addMagicProperty('screen', function ($el) {
            // bind resize event to window with debounce
            let update
            const updateScreen = () => {
                clearTimeout(update)
                update = setTimeout(() => {
                    $el.__x.updateElements($el)
                }, 150)

                Initialized = true
            }
            // bind resize event to window if not initialized
            if (!Initialized) {
                window.addEventListener('resize', updateScreen)
            }

....

```

muzafferdede replied:
i am trying to achieve this but still memory leaking.

```js
            let update

            window.addEventListener('resize', updateScreen => {

              clearTimeout(update)

              update = setTimeout(() => {
                  $el.__x.updateElements($el)
              }, 150)

              window.removeEventListener('resize', updateScreen, false)

            }, false)
```

HugoDF replied:
hmm what we're likely going to need to do is keep track of which components need to be updated.

something like:
```js
const screenComponents = [];
function debouncedScreensUpdate() {
    let update
    const updateScreens = () => {
        clearTimeout(update)
        update = setTimeout(() => {
            screenComponents.forEach(($el) => 
                $el.__x && $el.__x.updateElements($el)
            )
         }, 150)
     };
    return updateScreens;
}
let initialize = false;
const AlpineScreenMagicMethod = {
    start() {
        Alpine.addMagicProperty('screen', function ($el) {
            screenComponents.push($el);
            if (!initialized) {
                window.addEventListener('resize', debouncedScreensUpdate())
                initialized = true;
            }
```

HugoDF replied:
with something like the above, we can attach 1 resize handler but update all the components that use `$screen`.

Unless I misunderstand how `addMagicProperty` property callback works

HugoDF replied:
we could even move the `addEventListener` stuff to the `start()` function (outside of the `addMagicProperty` callback) so we don't even need to keep track of whether the listener is initialised or not

(`start()` should be called once vs the `addMagicProperty` callback which is called for every instance)

SimoTod replied:
That should work. Screencomponents could be a Set() rather than an array so we won't get duplications if you use the helper twice in the same component

HugoDF replied:
https://codepen.io/hugodf/pen/mdEZqaB?editors=1111 made this to clear up for me: `addMagicProperty` callback gets called on every `$screen` reference/invocation.

Here's a pared down version of "screen" working with the approach I mentioned above (1 resize listener, an array to keep track of elements to update) --> https://codepen.io/hugodf/pen/oNLroVB

HugoDF replied:
So probably what we can do in the end is:
```js
const screenComponents = [];
function debouncedScreensUpdate() {
    let update
    const updateScreens = () => {
        clearTimeout(update)
        update = setTimeout(() => {
            screenComponents.forEach(($el) => 
                $el.__x && $el.__x.updateElements($el)
            )
         }, 150)
     };
    return updateScreens;
}
let initialize = false;
const AlpineScreenMagicMethod = {
    start() {
        window.addEventListener('resize', debouncedScreensUpdate())
        Alpine.addMagicProperty('screen', function ($el) {
            screenComponents.push($el);
```

SimoTod replied:
We can do something like: Inside the foreach, if document doesn't contain $el, we remove $el from the list otherwise we trigger an update. This is also necessary for stuff like turbolinks.

KevinBatdorf commented on src/screen.js:
```diff
@@ -0,0 +1,65 @@
+                    ...config.get('breakpoints'),
+                    touch: 'ontouchstart' in window,
+                    portrait: height > width,
+                    landscape: width > height,
```
Wouldn't using `window.matchMedia("(orientation: portrait)").matches` be more of a native way to get this?

muzafferdede replied:
Oh, you probably didn't see 

> > About the second question, How would you implement landscape or touch in your helper?
> 
> ```js
>                 const screen = {
>                     ...config.get('breakpoints'),
>                     touch: 'ontouchstart' in window,
>                     portrait: height > width,
>                     landscape: width > height
>                 }
> ```
> 
> i already finish the $screen helper, now working on tests.
> 
> I have also tried to use matchMedia for `landscape` and `portrait` to check but to test it i have to mock matchMedia method which didn't work properly.
> If i could make it work, i would try to use media queries as well for breakpoints
> 
> i am trying to achieve this:
> 
> ```js
>                 const screen = {
>                     ...config.get('breakpoints'),
>                     touch: 'ontouchstart' in window,
>                     portrait: window.matchMedia("(orientation: portrait)").matches,
>                     landscape: window.matchMedia("(orientation: landscape)").matches
>                 }
> ```
> 
> Works on browser but i couldn't make it work on jest test. Specially don't know how to test `touch`. Maybe i just push the code so far and see if we can improve it.

Oh, probably you haven't read this. We also wanted that but can't test it. If you know how to test it, we can change it tho.

KevinBatdorf replied:
Ah ok. I read it at the time but forgot about it. I don't think we should change the implementation because we can't figure out how to test it though. I'll try and see what others do this week.

SimoTod replied:
We can test it but we have to mock match media (so a single test will cover landscape and portrait since there won't be any difference). We are mocking height and width anyway.
Out of curiosity, there is a pixel where the window is neither portrait nor landscape. It's worth investigating what mediaqueries do in this scenario and keep our implementation consistent if we go for a width/height comparison

KevinBatdorf replied:
Not sure about that, but I know Chrome lets you override the orientation in DevTools, including custom orientations.

https://developers.google.com/web/tools/chrome-devtools/device-mode/orientation#:~:text=The%20Sensors%20tab%20opens%20up,provide%20your%20own%20exact%20orientation.

That said, I just tested it now and `window.matchMedia` didn't seem to be affected by the change.

In Chrome, 1000x1000 was portrait, but not sure about other devices. It does feel like something we should let the device decide though if we can. Is there any reason not to use window.matchMedia other than the test? 

We could add in Cypress and start testing too.

muzafferdede replied:
Found a thing, working on it.

```js
                const screen = {
                    ...config.get('breakpoints'),
                    portrait: window.screen.orientation.angle === 90,
                    landscape: window.screen.orientation.angle === 0,
                }
```

so i can detect real orientation changes instead of comparing height and width. Mobile devices wont trigger resize event on screen orientation change, so better to do like so:
```js
        window.addEventListener('resize', debouncedScreensUpdate())
        window.addEventListener('orientationchange', debouncedScreensUpdate())
```

muzafferdede replied:
I think we are green now. We can test screen orientation in proper way now. 🎉

KevinBatdorf replied:
> Mobile devices wont trigger resize event on screen orientation change

Are you sure about this? We could probably add the listener anyway though.

`window.screen.orientation.angle` always gives `0` for me in Chrome

Another option is to just not include this part at all and instead let users add it themselves if they want it. We could just mention it in the docs.

muzafferdede replied:
Try chrome mobile view and change orientation there. You will see 90 differences

KevinBatdorf replied:
If Chrome always returns 0 when not in mobile mode, the code you're suggesting won't work properly. Both `portrait` and `landscape` could return true.

Personally I'm okay without this test and using `window.matchMedia("(orientation: portrait)").matches`, but if other's want it to be tested, and it's not feasible, then we should just let the user pass it in as an option. This seems like an edge case use anyway.

SimoTod replied:
Angle doesn't do what we want on desktop, it returns the orientation of the monitor (hardware) but the browser window could be resized to be in portrait mode, it would create confusion within users. 

If we use matchMedia there's really little to test, it's native so it does work. Mocking the object to test the condition passed in and return true and false, making sure the little code around preserve the right result would be enough for me. 

Breakpoints could use matchMedia too. We would need more discussion on this topic too so let's ignore it for now. 

I would say: let's get out a first version only supporting breakpoints which was the scope of the original PR. From there, let's open a discussion about orientation (same helper or different one? Listener we need, etc) and touch (possible $support helper). Would that work for you @muzafferdede?

muzafferdede replied:
Sure thing. I think except orientation, we have a solid breakpoints which as you said original scope of this pr. should i clean up the code and just push a breakpoints only $screen helper now?

SimoTod replied:
I think breakpoints only when you have time please. Thanks for keeping up with all the changes.

muzafferdede commented on src/screen.js:
```diff
@@ -0,0 +1,65 @@
+        let initialized
+
+        Alpine.addMagicProperty('screen', function ($el) {
+            // bind resize event to window with debounce
```
Yeah, we can change it. btw you sure all other helpers use arrow function? I saw few doesn't.

KevinBatdorf replied:
Just checked now and they all seem to be. Maybe you're thinking of the return function? Those appear to be mixed (and should probably be matched up in a PR too)

muzafferdede replied:
Hmm, i can't see a clear path, where i also copy one of the existing helper to start with so i just followed. So can i just use arrow functions instead of normal function anywhere i need to use?

muzafferdede replied:
Oh, i think my fork was older version where it was still used normal functions, after your merge the object pr, i see it.

muzafferdede commented on src/screen.js:
```diff
@@ -0,0 +1,65 @@
+                const screen = {
+                    ...config.get('breakpoints'),
+                    touch: 'ontouchstart' in window,
+                    portrait: height > width,
```
It is screen related. It is another way of checking if screen is on mobile device. Your perspective is probably from breakpoints so this not make sense to you a lot :D

KevinBatdorf replied:
Not only mobile devices support `touch` though. Screen sizes and position make sense but this doesn't seem to fit. 

Plus functionality wise it doesn't match with everything else here. Widths and orientations can change but `touch` support is a constant on the device.

muzafferdede replied:
```html
<div x-show="$screen('portrait'), $screen('lg') && $screen('touch')> 
   I need to be visible if screen is desktop size and support touch
</div>
```
Is this not make sense to you?

Consider uncommon screens. (I have worked on many different screen devices, there are weird cases where device orientation, size, etc. is not common)

KevinBatdorf replied:
I know _why_ you want it but I just don't think it fits here. You're checking whether a feature is supported instead of the current state of the browser width. Anyway, I don't necessarily feel strongly against it. Lets see what the others say.

muzafferdede replied:
Yeah sure, i am not strongly protecting it as well. I just think it's useful where it can detect it. Also IMO $screen helper should provide anything related to screen. 

For your reference [check this](https://www.npmjs.com/package/vue-screen#as-computed-properties)

KevinBatdorf replied:
That plugin isn't very popular (relative to other Vue plugins) so I don't think it's something to consider an authority on the topic.

But I think if you wanted to include all screen-related properties then I think that would work. Here's some more: https://developer.mozilla.org/en-US/docs/Web/API/Screen

SimoTod replied:
I like the idea of a support helper if there are multiple things to detect. E.g. You'll show a button only if geolocation is supported.

Other thing to investigate if we support orientation, does resize catches when you rotate a mobile phone or do we need a separate listener for it? (we don't need to find out today, i just posted the question so we won't forget :D ).

KevinBatdorf replied:
This does feel better:

```html
<div x-show="$screen('portrait') && $screen('lg') && $supports('touch')> 
   I need to be visible if screen is desktop size, portrait and supports touch
</div>
```

Just noticed your comment in the div there even says "supports", so you subconsciously agree :)

> Other thing to investigate if we support orientation, does resize catches when you rotate a mobile phone or do we need a separate listener for it? (we don't need to find out today, i just posted the question so we won't forget :D ).

I think this was an issue about 5 years ago but last I checked it was okay now and worked as expected and they deprecated other events.

muzafferdede replied:
Haha, well yeah, i am being stingy on  amount of the components. But yeah, if we can have other things stuff into $supports helper, that will be great 👍

muzafferdede replied:
And think about this:
```
window.AlpineMagicHelpers.Config = {
   supports: {
      *thing: *logic
   }
}}

where users can define their own thing and use it anywhere

KevinBatdorf replied:
Yeah well the supports helper can work similar to screen, but separate. 

Also, I'd be fine having hundreds of helpers. We would just eventually have to upgrade the documentation. I'd like to start implementing jQuery features more so I can imagine having a lot.

muzafferdede commented on src/screen.js:
```diff
@@ -0,0 +1,65 @@
+window.deferLoadingAlpine = function (callback) {
+    AlpineScreenMagicMethod.start()
+    alpine(callback)
+}
```
I am confused. Isn't there a config file which tells to editor about spacing and tabs? My editor's auto format is disabled. So not sure where is the spacing issues are at.

KevinBatdorf replied:
Line breaks. Check the other files for examples.

KevinBatdorf replied:
But like I mentioned, I'd prefer it the way you have it. I'd just prefer everything to look the same

muzafferdede commented:
Thanks @HugoDF @KevinBatdorf  @SimoTod. Let me know if anything else needs to be done. Other than that, i think it's ready to be merged.

HugoDF commented on src/screen.js:
```diff
@@ -0,0 +1,65 @@
+        clearTimeout(update)
+        update = setTimeout(() => {
+            screenComponents.forEach(($el) =>
+                $el && $el.__x && $el.__x.updateElements($el),
```
Did we want to remove falsy `$el` before doing the forEach @SimoTod ?

I think `$el`'s are the root elements of Alpine components so the only situation I can think where we might lose some is if using nested Alpine components inside of an `x-if` or something

SimoTod replied:
Are they going to be falsy or are they just going to be detached from the dom?
Other libraries (i.e. Turbolinks) may remove those elements from the Dom too. I think we should remove it from the list in that case.

HugoDF reviewed (approved):
Looks good to me 👍

SimoTod reviewed (approved):
I think it's solid enough to go in the next release now. 

We might want to reword the readme a bit for those not familiar with tailwind (do they really exist, lol?) and we also need to create examples/screen.html but those 2 points are not blockers and they can be tweaked on master before the release.

KevinBatdorf commented:
I think it looks good now. I agree the readme needs to be cleaned up a bit. Maybe also Update the CodePen a bit so it's a bit more interactive and clear (Maybe the active breakpoint are blue, for example). I can clean all that up tomorrow or someone else can as well.
